src/lavinmq/amqp/connection_factory.cr

```diff
@@ -47,7 +47,10 @@ module LavinMQ
         elsif proto != AMQP::PROTOCOL_START_0_9_1 && proto != AMQP::PROTOCOL_START_0_9
           socket.write AMQP::PROTOCOL_START_0_9_1.to_slice
           socket.flush
-          log.warn { "Unexpected protocol '#{String.new(proto.to_slice)}', closing socket" }
+          log.warn {
+            bad_protocol = String.new(proto.to_slice).gsub(/\n/, "\\n")
+            "Unexpected protocol '#{bad_protocol}', closing socket"
+            }
```

Comment:
only stringify the number of bytes we read, the rest will always be garbage. Then use `String#inspect` to render `\n` as `"\n"`, but also all other non non rendable characters, like `\t`, `\0` etc
```suggestion
          log.warn { "Unexpected protocol '#{String.new(proto.to_unsafe, count).inspect}', closing socket" }
```